src/signalrclient/connection_impl.cpp

```diff
@@ -375,20 +382,22 @@ namespace signalr
                 }
             });
 
-        std::thread([disconnect_cts, connect_request_done, connect_request_lock, callback, weak_connection]()
-        {
-            disconnect_cts->wait(5000);
-
+        timer(m_scheduler, [disconnect_cts, connect_request_done, connect_request_lock, callback](std::chrono::milliseconds duration) {
```

Comment:
What's the point of defining our own `cancellation_token` class if we have to poll it for cancellation? Let's add proper registration to cancellation_token instead of going through all this effort.

Reply:
The main problem is the 5 second timeout. Adding a callback to `cancellation_token` would let us run this callback immediately, but then how would we handle running the same callback after a 5 second timeout? By doing this polling, we solve both issues (with a small delay in the canceled case).